core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java

```diff
@@ -288,15 +288,27 @@ public List<BlockLocationInfo> getBlockLocations(URIStatus status)
         if (locations.isEmpty() && mFsContext.getPathConf(new AlluxioURI(status.getPath()))
             .getBoolean(PropertyKey.USER_UFS_BLOCK_LOCATION_ALL_FALLBACK_ENABLED)) {
           // Case 2: Fallback to add all workers to locations so some apps (Impala) won't panic.
-          locations.addAll(getHostWorkerMap().values());
-          Collections.shuffle(locations);
+          PropertyKey locKey = PropertyKey.USER_UFS_BLOCK_LOCATION_RETURN_LIMIT;
+          int count = mFsContext.getClusterConf().getInt(locKey);
+          if (count < 0) {
+            throw new IllegalArgumentException("Property" + locKey.getName()
+                + " should not be set to a negative number");
+          }
+          List<WorkerNetAddress> addresses = getShuffleWorkerAddressList();
+          locations.addAll(addresses.subList(0, Math.min(addresses.size(), count)));
```

Comment:
this line has many copies, you can just write an easy for loop
```
for (i = 0; i < count; i++) {
  locations.add(addresses.get(i));
}
```

Reply:
done.